erpnext/stock/doctype/packed_item/packed_item.py

```diff
@@ -63,38 +69,65 @@ def update_packing_list_item(doc, packing_item_code, qty, main_item_row, descrip
 	bin = get_bin_qty(packing_item_code, pi.warehouse)
 	pi.actual_qty = flt(bin.get("actual_qty"))
 	pi.projected_qty = flt(bin.get("projected_qty"))
+	packed_items_list.append(pi.idx)
 
-def make_packing_list(doc):
+def make_packing_list(doc, doctype=None, name=None, fieldname=None):
 	"""make packing list for Product Bundle item"""
 
 	if doc.get("_action") and doc._action == "update_after_submit": return
 
+	maintain_packed_items_list = 0
+	if doctype is not None and doctype == "Sales Order":
+		maintain_packed_items_list = cint(frappe.get_value(doctype, name, fieldname))
+	else:
+		if hasattr(doc, 'items'):
+			if hasattr(doc.get("items")[0], 'against_sales_order'):
+				maintain_packed_items_list = cint(frappe.get_value("Sales Order", doc.get("items")[0].against_sales_order,
+				                                                   "maintain_packed_items_list"))
+
+	if maintain_packed_items_list and doc.doctype == "Sales Order" and doc.docstatus != 0:
+		return
+
 	parent_items = []
+	packed_items_list = []
+
 	for d in doc.get("items"):
 		if frappe.db.get_value("Product Bundle", {"new_item_code": d.item_code}):
-			for i in get_product_bundle_items(d.item_code):
-				update_packing_list_item(doc, i.item_code, flt(i.qty)*flt(d.stock_qty), d, i.description)
+			if maintain_packed_items_list and doc.doctype in ["Sales Invoice", "Delivery Note"] and d.so_detail is not None:
+				for i in get_current_product_bundle_items(d.item_code, d.so_detail):
+					update_packing_list_item(doc, i.item_code, flt(i.qty), d, i.description, packed_items_list)
+			else:
+				for i in get_product_bundle_items(d.item_code):
+					update_packing_list_item(doc, i.item_code, flt(i.qty)*flt(d.stock_qty), d, i.description, packed_items_list)
 
 			if [d.item_code, d.name] not in parent_items:
 				parent_items.append([d.item_code, d.name])
 
-	cleanup_packing_list(doc, parent_items)
+	cleanup_packing_list(doc, parent_items, packed_items_list, maintain_packed_items_list)
 
-def cleanup_packing_list(doc, parent_items):
+def cleanup_packing_list(doc, parent_items, packed_items_list, maintain_packed_items_list):
 	"""Remove all those child items which are no longer present in main item table"""
 	delete_list = []
 	for d in doc.get("packed_items"):
-		if [d.parent_item, d.parent_detail_docname] not in parent_items:
-			# mark for deletion from doclist
-			delete_list.append(d)
+		if maintain_packed_items_list:
+			if d.idx not in packed_items_list:
+				# mark for deletion from doclist
+				delete_list.append(d)
+		else:
+			if [d.parent_item, d.parent_detail_docname] not in parent_items:
+				# mark for deletion from doclist
+				delete_list.append(d)
 
 	if not delete_list:
 		return doc
 
 	packed_items = doc.get("packed_items")
 	doc.set("packed_items", [])
+	c = 1
```

Comment:
Can you please add a comment, or use a better variable name to note why you are doing this step exactly?